Fixed #21803 -- Added support for post-commit hooks

pelme opened the pull request:
This PR is an attempt to introduce django-transaction-hooks into Django.

Ticket: https://code.djangoproject.com/ticket/21803
Recent discussion: https://groups.google.com/forum/#!topic/django-developers/qWTL40W4Erw

Todo:

* [X] Port existing test suite into Django's test suite
* [X] Morph `TransactionHooksDatabaseWrapperMixin` into Django's `BaseDatabaseWrapper`
* [x] Fix the issue described in https://github.com/carljm/django-transaction-hooks/issues/10
* [x] Port/adapt docs

carljm commented on tests/transaction_hooks/tests.py:
```diff
@@ -0,0 +1,223 @@
+    # ``connection.autocommit = True`` directly instead), so we have to run
+    # hooks immediately on commit instead of waiting for autocommit to be
+    # restored. If a hook tries to create an internal transaction of its own,
+    # this fails with an error. Need a better solution here; could be fixed if
```
So this is an issue I'd like input on from people who know the ORM and transaction code (paging @aaugustin @akaariai and other people whose names start with A :-) ).

After committing or rolling back a transaction, `Atomic` calls `connection.set_autocommit(True)` on all database backends except SQLite (technically, all databases except those with `features.autocommits_when_autocommit_is_off`, which means SQLite). On SQLite `connection.set_autocommit(True)` is never called; instead the connection's attribute is just set directly: `connection.autocommit = True`. Because we can't hook into the latter (unless we make it a property), on SQLite we can't delay running the commit hooks until after autocommit has been turned back on, which in turn means that you can't open a transaction inside an `on_commit` hook on SQLite without Django squawking at you ("Your database backend doesn't behave properly when autocommit is off. Turn it on before using 'atomic'." - ironically in this situation autocommit is in fact on, since `Atomic` never turns it off on SQLite and the transaction has ended, but `Atomic` hasn't yet gotten around to telling the connection that it's back in autocommit mode).

We could just accept that behavior, but I think it will cause problems for people using SQLite in dev who have `on_commit` hooks that open their own transaction (this is perhaps an odd thing to do, but I'm sure some people will do it.) So I think we should fix it. Some options I see for a fix:

1) Instead of having `Atomic` poke directly at the `connection.autocommit` flag on a SQLite backend, have the backend manage it internally. That is, have the `_start_transaction_under_autocommit()` method set it to `False`, and have both `commit()` and `rollback()` set it back to `True`, only if it was originally set to `False` by entering a transaction while still in autocommit. Con: this would require a new flag to track whether autocommit is really off, or just off-because-we-started-a-transaction. Pro: this would ensure that the `autocommit` flag more accurately reflects the current state of autocommit on SQLite, eliminating the period of inaccuracy between commit/rollback and `Atomic` getting round to setting it back to `True` that causes this problem.

2) Override `set_autocommit` on the SQLite backend so that it does nothing but change the value of the flag. Pro: `Atomic` could treat all backends equally and stop worrying about `features.autocommits_when_autocommit_is_off`. Con: this is backwards-incompatible for anyone who is actually intentionally turning autocommit off on SQLite (is there any good reason why anyone would do this, given that it's broken anyway?).

3) Add a totally new API on the connection object that `Atomic` calls once a transaction is fully finished (committed or rolled back) and the connection has been fully reset to its post-transaction state (autocommit on). Something like `connection.transaction_finished()`. Then the transaction-hooks would run in this method. Pro: Doesn't require otherwise messing with the transaction/autocommit logic. Con: Adds new API of perhaps-questionable semantic value. Ties the transaction hooks even more closely to use of `Atomic` (or some other smart higher-level transaction API that knows about the need to call this method) rather than to the underlying transaction-state methods on the connection.

I think these options are listed roughly in order of my preference (though I don't have strong feelings between (1) and (2), and could be OK with any of them). It's quite likely that I'm missing some obvious reason why one or more of these wouldn't work, though. So - your thoughts?

carljm replied:
Oh, there is a fourth option: turn the `autocommit` flag on the SQLite backend into a property, so that we can hook into `connection.autocommit = True` to run the transaction hooks.

carljm commented:
@pelme This is great! Thanks so much for working on it, much appreciated.

I just wrote a long treatise on one of the two remaining todo items; curious to hear what you and others think.

Besides that issue and docs, I think this looks ready to go!

pelme commented:
Thanks for the quick and detailed feedback, Carl! I'll start porting/writing the docs.

I'm not deep enough into the transaction internals and to bring anything useful to the discussion about the SQLite issue. However, I would be happy to provide an implementation for when any option when it is decided upon (if it actually is helpful and makes things easier for you :-)).

aaugustin commented:
@carljm I'd like to suggest a variant of option 2: move the workaround for https://bugs.python.org/issue10740 from `Atomic` to the database backend.

(Pro-tip: you can find that ticket by googling for "python sqlite3 transaction bug"... sigh...)

This requires replacing the following two blocks of code by a database backend API:

            if connection.features.autocommits_when_autocommit_is_off:
                connection._start_transaction_under_autocommit()
                connection.autocommit = False
            else:
                connection.set_autocommit(False)

and:

                if connection.features.autocommits_when_autocommit_is_off:
                    connection.autocommit = True
                else:
                    connection.set_autocommit(True)

For example we could add a keyword argument to `set_autocommit`:

    def set_autocommit(self, autocommit, cheat_if_backend_autocommits_when_autocommit_is_off=False):
        if cheat_if_backend_autocommits_when_autocommit_is_off and self.features.autocommits_when_autocommit_is_off:
            if autocommit:
                self._start_transaction_under_autocommit() 
            self.autocommit = autocommit
        else:
            # use the current implementation.

I believe that encapsulate the ugliness a bit better...

`Atomic` has another reference to `autocommits_when_autocommit_is_off`:

            if not connection.get_autocommit():
                # Some database adapters (namely sqlite3) don't handle
                # transactions and savepoints properly when autocommit is off.
                # Turning autocommit back on isn't an option; it would trigger
                # a premature commit. Give up if that happens.
                if connection.features.autocommits_when_autocommit_is_off:
                    raise TransactionManagementError(
                        "Your database backend doesn't behave properly when "
                        "autocommit is off. Turn it on before using 'atomic'.")

I'm not sure it's possible to move this safely into the call to `connection.set_autocommit(False)` that happens 40 lines below because flags on the connection state will have changed, and I'm not sure what happens then.

pelme commented:
I have ported/adapted the docs from django-transaction-hooks and added a release note.

I will work on @aaugustin's suggested approach tomorrow, thanks a lot for the detailed discussions and explanations!

aaugustin commented:
@pelme I think the idea is worth trying, but if it doesn't work well, don't waste too much time, we'll go back to the drawing board :-) Thanks!

pelme commented:
Ok, I've implemented option 2 with Aymerics addition in cbe38840498bd946d79ea240cb49850e38f4d38b. The following commit (d2a0ec07b6011f4774ceba0f1283073a0bd22fcd) makes the commit hooks always run from `set_autocommit(False)`.

Basically `set_autocommit(False)` now accepts the `cheat_if_backend_autocommits_when_autocommit_is_off` that forces a BEGIN with an explicit transaction. This solves the con of making this change backward incompatible, it is now opt-in to have `connection.set_autocommit(False)` to not be broken for those that somehow relied on the old behavior.

Atomic just uses `set_autocommit()` instead of modifying `connection.autocommit` directly, which then makes it possible to have the commit hooks always called from within `connection.set_autocommit(True)`

The `connection.features.autocommits_when_autocommit_is_off` check when entering an atomic block is still left in `Atomic`, does that cause any problems? I am not sure if there is a problem that it just remains as is?

Tests pass locally for me on SQLite and Postgres and this seems to work. Needless to say, this is very complicated and I have not fully grasped this and very much like you to look properly on this. :-)

pelme commented:
Other thoughts:

1) Should we provide a `django.db.transaction.on_commit(func, using=None)` function to match the other transaction API+promote that function in the transaction docs?

2) Does it makes sense to allow on_commit hooks when being in manual transaction mode? Surely this doesn't make sense and would lead to weird surprises? (Currently fn is called directly)

```python
connection.set_autocommit(False)
connection.on_commit(fn)
connection.commit()
```

aaugustin commented:
`cheat_if_backend_autocommits_when_autocommit_is_off` was a bit tongue-in-cheek; perhaps we can find a better name :-)

Otherwise, this looks quite reasonable to me.

pelme commented:
Agreed, my mind was kind of busy with other things when I copy-pasted that :) I'll try to think of a better name and commit it later!

(Should I be worried about the docs build failure? http://djangoci.com/job/docs/2209/console reports a bunch of errors in the release notes that does not seem to be related to this issue, right?)

pelme commented:
* dd358a1 renames the new `set_autocommit` argument to `force_begin_transaction_with_broken_autocommit`. I am still not super happy about that name, suggestions welcome.

* I added `on_commit` to `django.db.transaction` and updated docs/tests to make it more consistent with other transaction APIs

* `on_commit` registrations when in non-autocommit mode and outside of atomic blocks now raises an exception. The previous behavior was confusing. We could fix this, but I don't see much point in this either -- if you're rolling your own transaction handling, I guess `on_commit` is not very useful anyways? This can also be added later of someone is interested in it.

I would be happy for any kind of feedback of the above changes/anything else needed to get this ready for merge.

aaugustin commented:
Regarding the parameter name, what about `explicit_begin`?

pelme commented:
I thought about something like that, but I figured it might be confusing to the other database backends that does not suffer from this issue since they never issue a BEGIN like that. I.e calling `set_autocommit(False, explicit_begin=True)` on Postgres may have you think that you'd actually get an explicit BEGIN. Therefore I added the `broken_autocommit` bit to it. Maybe `explicit_begin_for_bad_autocommit`? A bit shorter but still long and ugly :/

(I'm fine either way, let me know what you think and I'll just change it)

aaugustin commented:
I gave it some thought and circled back to `force_begin_transaction_with_broken_autocommit`... Your original proposal is good enough!

carljm commented:
Hi all - I was out for the weekend, but this all looks great. I think Aymeric's API suggestion is better than any of the ones I proposed, I don't have any ideas for a better parameter name than `force_begin_transaction_with_broken_autocommit` (it's long but clear), I agree that it's best to just raise an error if `on_commit` is used when not in autocommit mode (the transaction-hooks docs already strongly warn against this, but I never got around to actually making it an error), and I always had it in mind with transaction-hooks that the connection API might just be the low-level API and should be wrapped by a function in `django.db.transaction` to parallel all the other transaction APIs.

carljm commented on docs/releases/1.9.txt:
```diff
@@ -25,7 +25,18 @@ Python 3.2 and added support for Python 3.5.
 ========================
 
-...
+Ability to perform actions after a transaction commit
```
I think 'Performing actions after a transaction commit` reads less awkwardly

carljm commented on docs/releases/1.9.txt:
```diff
@@ -25,7 +25,18 @@ Python 3.2 and added support for Python 3.5.
+
+A new hook, :func:`~django.db.transaction.on_commit` has been introduced to make
+it possible to perform actions after a database transaction is successfully
+committed. This is useful when sending emails, creating Celery tasks or invalidate
```
"This is useful for tasks such as sending notification emails, creating queued tasks, or invalidating caches."

Current sentence has a grammatical error with "invalidate", and "useful when" suggests that this is an exclusive list. Also I'm not sure there's any reason to specifically name Celery here (I know, I did in the django-transaction-hooks docs), there are other task queues that people use.

carljm commented on django/db/backends/base/base.py:
```diff
@@ -501,3 +541,22 @@ def schema_editor(self, *args, **kwargs):
+    def on_commit(self, func):
+        if self.in_atomic_block:
+            # transaction in progress; save for execution on commit
+            self.run_on_commit.append((self.savepoint_ids[:], func))
```
This is a very minor thing, but I wonder why I didn't make the savepoint-ids item in this tuple a set instead of a list, considering its only used for containment checks? I think it would be a trivial change, just change this from `self.savepoint_ids[:]` to `set(self.savepoint_ids)`

aaugustin replied:
This will become a performance issue for the database before it becomes one for the Python process :-)

carljm replied:
Yes, that's true. :-)

carljm commented on django/db/backends/base/base.py:
```diff
@@ -71,6 +71,15 @@ def __init__(self, settings_dict, alias=DEFAULT_DB_ALIAS,
         self._thread_ident = thread.get_ident()
 
+        # a list of no-argument functions to run when the transaction commits;
+        # each entry is an (sids, func) tuple, where sids is a list of the
```
if the `sids` element in this tuple is changed from a list to a set, this comment would also need to be updated accordingly

carljm commented on docs/topics/db/transactions.txt:
```diff
@@ -252,6 +252,130 @@ by Django or by third-party libraries. Thus, this is best used in situations
+.. versionadded:: 1.9
+
+Sometimes you need to fire off an action related to the current database
+transaction, but only if the transaction successfully commits. Examples: a
```
The sentence fragment here was lazy writing on my part. Better would be "Examples might include a..."

I also might lean towards taking out the specific reference to Celery here, but I don't feel strongly either way.

carljm replied:
Eh, never mind re Celery. It's used several other places as an example, and I think that's fine.

carljm commented on docs/topics/db/transactions.txt:
```diff
@@ -252,6 +252,130 @@ by Django or by third-party libraries. Thus, this is best used in situations
+    you set disable autocommit on your connection and/or use lower-level
+    transaction APIs directly, the :func:`on_commit` hook will not work.
+
+Commit hooks are not run until autocommit is restored on the connection
```
I wonder if these two paragraphs would belong better added to the "timing of execution" section above?

carljm commented on docs/topics/db/transactions.txt:
```diff
@@ -252,6 +252,130 @@ by Django or by third-party libraries. Thus, this is best used in situations
+-------------------------------------
+
+.. warning::
+    On-commit hooks only works correctly in :ref:`autocommit
```
I'm not sure that this paragraph really needs to be called out as a warning block, especially since trying to do what it cautions against will result in a clear error anyway, not silent confusing behavior.

There are also some subject-verb agreement issues in this sentence. Suggested rewrite:

> On-commit hooks only work with autocommit mode and the atomic (or ATOMIC_REQUESTS) transaction API. Calling on_commit when autocommit is disabled and you are not within an atomic block will result in an error.

carljm commented on docs/topics/db/transactions.txt:
```diff
@@ -252,6 +252,130 @@ by Django or by third-party libraries. Thus, this is best used in situations
+---------------------
+
+A rollback hook is harder to implement robustly than a commit hook, since a
+variety of things can cause an implicit rollback. For instance, your database
```
More lazy writing on my part. Change to "For instance, if your database connection is dropped because your process was killed without a chance to shut down gracefully, your rollback hook will never run."

carljm commented:
Made a few small comments on docs. Once those are fixed, as far as I can see this is ready to be squashed down to a single commit and merged.

pelme commented:
Thanks a lot for the comments, I will try to fix them all tonight!

(Should I squash+force push to this PR when the changes are done, or should I create a separate PR for the final commit?)

carljm commented:
@pelme Just squash and force push to this PR. Though you could also wait a bit and see if @timgraham has a chance to review first, he often catches details I miss :-)

pelme commented:
Thanks, I'll fix everything and wait with the squashing until it all looks good :)

timgraham commented on django/db/backends/base/base.py:
```diff
@@ -253,6 +269,10 @@ def savepoint_rollback(self, sid):
         self._savepoint_rollback(sid)
 
+        # remove any callbacks registered while this savepoint was active
+        self.run_on_commit = list(filter(
```
A list comprehension may be more readable/Pythonic.

timgraham commented on docs/topics/db/transactions.txt:
```diff
@@ -252,6 +252,130 @@ by Django or by third-party libraries. Thus, this is best used in situations
+    from django.db import transaction
+
+    def do_something():
+        pass  # send a mail, invalidate a cache, fire off a Celery task, what-have-you.
```
what-have-you -> etc.

timgraham commented on docs/topics/db/transactions.txt:
```diff
@@ -252,6 +252,130 @@ by Django or by third-party libraries. Thus, this is best used in situations
+unhandled exception is raised in an :func:`atomic` block), your function will
+be discarded and never called.
+
+
```
one blank line between sections please

timgraham commented on docs/topics/db/transactions.txt:
```diff
@@ -252,6 +252,130 @@ by Django or by third-party libraries. Thus, this is best used in situations
+
+The function you pass in will be called immediately after a hypothetical
+database write made at the same point in your code is successfully
+committed.
```
this fits on the previous line

timgraham commented on tests/transaction_hooks/tests.py:
```diff
@@ -0,0 +1,220 @@
+
+    Creation/checking of database objects in parallel with callback tracking is
+    to verify that the behavior of the two match in all tested cases.
+
```
no blank line needed

timgraham commented:
I'll give the docs another look after you incorporate Carl's feedback.

pelme commented:
@timgraham I have updated the PR with all comments so far, could you please take a final look at the docs? :-)

(The CI build seems to have some problems, but I think the tests should be fine)

carljm commented:
@timgraham Did you want to take a last look at the docs here before I ask @pelme to rebase and squash?

timgraham commented:
I'll do it tomorrow if nothing else distracts me.

timgraham commented on django/db/backends/base/base.py:
```diff
@@ -286,15 +307,35 @@ def get_autocommit(self):
         return self.autocommit
 
-    def set_autocommit(self, autocommit):
+    def set_autocommit(self, autocommit, force_begin_transaction_with_broken_autocommit=False):
```
`force_begin_transaction_with_broken_autocommit` is a really long name, but its purposes still isn't obvious to me. Could you add some info to the docstring about its purpose? If the name could be simplified/shortened, I think that would be beneficial as well.

carljm replied:
It's explained in a comment within the body of the method, but having an explanation in the docstring (also or instead) would be good.

Neither myself nor @aaugustin nor @pelme could come up with a shorter name that seemed accurate, but ideas are welcome.

pelme replied:
I've updated the docstring and changed the text a bit with . I still haven't figured out a better name for the parameter or how to make it obvious. It is working around an non-obvious bug, it took me quite a while just to understand what was going on... :/

timgraham commented on docs/topics/db/transactions.txt:
```diff
@@ -252,6 +252,126 @@ by Django or by third-party libraries. Thus, this is best used in situations
+
+    transaction.on_commit(lambda: some_celery_task.delay('arg1'))
+
+The function you pass in will be called immediately after a hypothetical
```
This sentence is confusing to me. An example might help.

carljm replied:
Yeah, an example or two would probably help. Simple examples could be drawn from the tests, maybe?

I also think that the section below labeled "Savepoints" should maybe be moved up into this section, right after the paragraph about rollbacks? It's all really discussing the same issue: under what conditions will your callback be called, or not called?

timgraham commented on docs/topics/db/transactions.txt:
```diff
@@ -252,6 +252,126 @@ by Django or by third-party libraries. Thus, this is best used in situations
+etc), this should be fine. If it's not (if your follow-up action is so critical
+that its failure should mean the failure of the transaction itself), then you
+don't want to use the :func:`on_commit` hook.  (Instead, you may want
+`two-phase commit`_.)
```
I think it would be better if we had a Django example for "two-phase commit" instead of linking to the wikipedia article. Is it too complicated?

carljm replied:
I haven't seen anyone do it in Django, and Google doesn't turn anything up that I can find. It's certainly not easy, and would require using underlying methods on the DBAPI connection; Django doesn't offer any higher-level support for it. We could link to something like http://initd.org/psycopg/docs/usage.html#tpc (that's Postgres-specific, but I'm not sure which other supported Django databases support two-phase commit?) And/or we could link to https://www.python.org/dev/peps/pep-0249/#optional-two-phase-commit-extensions in the Python DBAPI spec.

timgraham commented on tests/transaction_hooks/tests.py:
```diff
@@ -0,0 +1,219 @@
+        Thing.objects.create(num=num)
+        transaction.on_commit(lambda: self.notify(num))
+
+    def assert_done(self, nums):
```
change assertions to use camel case?

timgraham commented:
Here are [suggested edits](http://dpaste.com/1FZXSSZ). One point to clarify: in my mind, the `on_commit()` function is considered the "hook", and the functions that are registered with it are referred to as "callbacks" or "functions" (not "hooks"). The docs mixed this up a bit and my edits cleaned it up, but let me know if this was in error.

carljm commented on docs/topics/db/transactions.txt:
```diff
@@ -252,6 +252,126 @@ by Django or by third-party libraries. Thus, this is best used in situations
+The function you pass in will be called immediately after a hypothetical
+database write made at the same point in your code is successfully committed.
+
+If you register a callback while there is no transaction active, it will be
```
I think this paragraph would flow better after the following paragraph (about rollback). Commit and rollback are the primary cases to discuss; using `on_commit` outside of a transaction is an edge case.

carljm commented on docs/topics/db/transactions.txt:
```diff
@@ -252,6 +252,126 @@ by Django or by third-party libraries. Thus, this is best used in situations
+If you register a callback while there is no transaction active, it will be
+executed immediately.
+
+If that hypothetical database write is instead rolled back (typically when an
```
I'm not sure the parenthetical here adds much value. The point of this paragraph is that if a rollback occurs, the callback won't be called; discussing the possible causes of a rollback is an out-of-scope distraction.

carljm commented:
Thanks @timgraham - I like all your suggested edits. I think you're right about the correct usage of "hook" - that was laziness when I wrote the transaction-hooks docs (in fact the wrong usage is even in the name of the package.)

pelme commented:
Thanks for the comments, I will try to update/comment as soon as possible!

timgraham commented:
I think it's time to rebase and squash commits.

pelme commented:
Huge thanks to you for being so responsive and giving great feedback!

I was not able to follow up all comments yet, the Github UI hides some comments that I have not yet fixed because it changes other unrelated wording. I will go through everything and make sure all changes is fixed and push a rebased commit.

pelme commented:
I've addressed the remaining bits by moving the savepoint section (c236b26a4ec3f6c874b7a66fd814f97fbc8333f5) and added examples on how savepoints/rollbacks are handled (72ddaa0f2366af60653f0b3fc51d1b1f4e2d29b0). I've also added links to psycopg and DB-API with regard to 2PC in 570cc92def4e6c8771398fcd7dc337f8f228d5dc.

It has been rebased+squashed on top of the latest master, all CI checks seems good. I think it should be good to go. :)

/cc @timgraham

timgraham commented on docs/topics/db/transactions.txt:
```diff
@@ -252,6 +252,149 @@ by Django or by third-party libraries. Thus, this is best used in situations
+    transaction.on_commit(lambda: some_celery_task.delay('arg1'))
+
+The function you pass in will be called immediately after a hypothetical
+database write made at the same point in your code is successfully committed.
```
Is this phrase correct? I'm having trouble understanding it.

pelme replied:
Sorry, I think that sentence was discussed earlier. I'll reword it and push a fix!

timgraham replied:
Great, please apply this patch too: http://dpaste.com/09M4WH9

pelme commented:
Patch from @timgraham is applied in commit 1891876de5677f6b126a739051d724db609b1d59.

Updates for the confusing section on when a transaction is committed is fixed in f620eae1e010d9a1f3838f622d20c88542700009.

I have squashed and force pushed a new commit that incorporates those changes!

timgraham commented:
merged in 00a1d4d042a7afd139316982c9b57e87d26a894f, thanks!
